### ISSUE TITLE
Performance improvement at login

### ISSUE THREAD

**Issue by mgriner** [arkayenro/arkinventory#1027]
In the process of trying to optimize the loading time at login for my search plugin I came across a performance improvement you may be able to implement very easily.

Basically, in ArkInventoryTranslate.lua and ArkInventoryRules.lua calls are made like this:

```lua
ArkInventory.Frame_Main_Generate( nil, ArkInventory.Const.Window.Draw.Recalculate )
```

This function then sets the draw status and then calls the actual draw functions for every location.  This causes a pretty big stutter at login when translation/rules are loaded,  literally **3-4 seconds of unplayable FPS** for me.  At first I thought maybe I could get away with not calling Frame_Main_Generate at all.  This seemed like it worked fine, player logs in, no stutter, open bags everything looked great.  After some testing though I realized Frame_Main_Generate still needs to be called but it can skip a bunch of locations and everything will still work without the stuttering.

Here is the solution I came up with, there is an (overly) detailed explanation of why I think this works below the code.  My assumptions might be wrong but it seems to work well.  I also don't know if other calls to Frame_Main_Generate with location = nil to process all locations might be causing performance issues elsewhere, I only checked issues at login.  Also, it may be the case that a better solution might be to optimize the Frame_Main_DrawLocation function directly.

**Proposed Solution:**

Add an optional parameter and a new if block to `Frame_Main_Generate` in `ArkInventory.lua`:

```lua
function ArkInventory.Frame_Main_Generate( location, drawstatus, isStartupCall )
   for loc_id, loc_data in pairs( ArkInventory.Global.Location ) do
      if loc_data.canView and ( not location or loc_id == location ) then
         if (not isStartupCall) or (isStartupCall and ArkInventory.Global.Location[loc_id].drawState ~= 0) then
            ArkInventory.Frame_Main_DrawStatus( loc_id, drawstatus )
            ArkInventory.Frame_Main_DrawLocation( loc_id )
         end
      end
   end
end
```

Adjust the calls in `ArkInventoryTranslate.lua (line 363)` and `ArkInventoryRules.lua (line 112)` to include the optional flag indicating this special situation.

```lua
ArkInventory.Frame_Main_Generate( nil, ArkInventory.Const.Window.Draw.Recalculate, true )
```

## **tldr explanation:**
**Ignore drawing locations at login unless they have been prematurely opened and therefore require a recalculate, instead defer the initial sort/draw to first time location is opened.**

## **long detailed explanation:**

After some testing I realized one caveat about just removing the Frame_Main_Generate calls.  If you open your bags up immediately after logging in they are not sorted/organized yet.  That call to Frame_Main_Generate was sorting/organizing bags that were already open once the translation/rules have been loaded.  Removing that call results in bags remaining unsorted/unorganized until the user closes said bag and then re-opens it.  However, if you do NOT open your bags immediately but wait a little bit then open them all is good.

What I think is happening is something like this: bags start with a state 0 (init).  If they are opened and their state is 0 the bag is going to sort/organize itself and set its own state to 4 (done).  But if the translation/rules have not been loaded yet it does not sort/organize itself, but still sets its own state to 4 (done).

The call to Frame_Main_Generate was setting the state to 1 (recalculate) which also causes the bags to sort/organize.

So here is a simplified view of what is currently happening:
Log in->All bag states = 0->Translation/Rules Loaded->All Bag states set to 1->All bags drawn, all states set to 4->STUTTER->Bags are now sorted/organized when opened

OR if opening bags immediately

Log in->All bag states = 0->Bag opened early->Bag drawn but not sorted, state set to 4->Translation/Rules Loaded->All bag states set to 1 (including the opened bag)->STUTTER->Bags are now sorted/organized when opened

Removing Frame_Main_Generate call results in the following:
Log in->All bag states = 0->Translation/Rules Loaded->Bags are sorted/organized on-demand when opened for the first time (no stutter at login)

OR if opening bags immediately
Log in->All bag states = 0->Bag opened early->Bag drawn but not sorted, state set to 4->Translation/Rules Loaded->All bags are sorted/organized on-demand when opened for the first time BUT ALREADY OPENED BAGS REQUIRE CLOSE/REOPEN (no stutter at login)

This lead me to the solution listed at the top of the post.  I just add an optional flag to the Frame_Main_Generate function to indicate when it is one of these special cases.  Then, inside Frame_Main_Generate I just add an if statement in the for loop that checks to see if it is one of these special cases, if it is and the bag state does not equal 0 (meaning the bag has already been opened) we set the state to 1 (recalculate) and draw.  Otherwise, just ignore that location, it will get sorted/drawn on its first open event.

This gets rid of the lag at login by not sorting/drawing every location but also makes sure that if the user opens bags before everything is ready those bags will get sorted once the translation/rules are loaded.

**Comment by arkayenro** [comment]
there is old code in the Frame_Main_Draw function to that i added to handle an issue with the Init state and threading, which i think i also fixed another way so it shouldnt be there any more

line 4463: if ArkInventory.Global.Location[loc_id].drawState > ArkInventory.Const.Window.Draw.Init and not frame:IsVisible( ) and not ArkInventory.Global.Location[loc_id].show then

should just be

if not frame:IsVisible( ) and not ArkInventory.Global.Location[loc_id].show then

try that in your copy and see if it goes away

**Comment by mgriner** [comment]
Yep, that change at 4463 fixes it. 😃

**Comment by arkayenro** [comment]
it'll be in the BFA alphas/releases
